launcher/src/main/java/org/wso2/carbon/launcher/extensions/DropinsBundleDeployerUtils.java

```diff
@@ -0,0 +1,273 @@
+                            String bundleVersion = manifest.getMainAttributes().getValue("Bundle-Version");
+
+                            if (bundleSymbolicName == null || bundleVersion == null) {
+                                throw new IOException(
+                                        "Required bundle manifest headers do not exist in " + jarFile.toString());
+                            } else {
+                                if (bundleSymbolicName.contains(";")) {
+                                    bundleSymbolicName = bundleSymbolicName.split(";")[0];
+                                }
+                            }
+
+                            //  checks whether this bundle is a fragment or not
+                            boolean isFragment = (manifest.getMainAttributes().getValue("Fragment-Host") != null);
+                            int defaultBundleStartLevel = 4;
+                            BundleInfo generated = new BundleInfo(bundleSymbolicName, bundleVersion,
+                                    "../../" + dropinsDirectory + "/" + fileName, defaultBundleStartLevel, isFragment);
+                            return Optional.of(generated);
+                        }
+                    }
+                } else {
+                    return Optional.empty();
+                }
+            }
+        } else {
+            throw new IOException("Invalid OSGi bundle path: " + bundlePath);
+        }
+    }
+
+    /**
+     * Returns true if the specified bundles.info file requires to be updated, else false.
+     * <p>
+     * The OSGi bundle information specified are compared with the OSGi bundle information specified in the
+     * existing bundles.info file, specified. If the OSGi bundle information of dropins bundles are matching,
+     * there is no requirement to update the bundles.info file, again.
+     *
+     * @param newBundleInfo          the new OSGi bundle information
+     * @param existingBundleInfoFile the existing bundles.info file to be checked
+     * @return true if the specified bundles.info file requires to be updated, else false
+     * @throws IOException if an I/O error occurs
+     */
+    public static boolean hasToUpdateBundlesInfoFile(List<BundleInfo> newBundleInfo, Path existingBundleInfoFile)
+            throws IOException {
+        if ((existingBundleInfoFile != null) && (Files.exists(existingBundleInfoFile))) {
+            List<BundleInfo> existingBundlesInfo = new ArrayList<>();
+            Files.readAllLines(existingBundleInfoFile).stream().filter(line -> !line.startsWith("#")).
+                    map(BundleInfo::getInstance).
+                    filter(BundleInfo::isFromDropins).forEach(existingBundlesInfo::add);
+
+            long newBundleInfoCount = Optional.ofNullable(newBundleInfo).orElse(new ArrayList<>()).size();
+            if (existingBundlesInfo.size() == newBundleInfoCount) {
+                long nonMatchingBundleInfoCount = Optional.ofNullable(newBundleInfo).
```

Comment:
Please have a look at this class: https://github.com/wso2/carbon-kernel/blob/5c2f48bfcaa8e12fb22c5bc5a624d8bafe5f9d66/core/src/main/java/org/wso2/carbon/kernel/internal/startupresolver/StartupComponentManager.java#L202